### PR TITLE
Trainercode Feature

### PR DESCRIPTION
**Pull request by studplus**
Stores Trainercode to DB
Shows Trainercode in Raidpoll and Alarm (raidbot message)
Works only when anytime button has been pushed

### CODE REVIEW

**Comment by Artanicus** [comment]
I'm assuming tying this to anytime is just a stopgap until we have the full feature for invite begging?

**Review by Artanicus — CHANGES_REQUESTED** [review]
A few notes of parts that were clearly for the trainercode. Can you cherrypick the trainercode features to a separate branch that doesn't include the trainername changes? Or are the changes too interlinked to make that feasible?

File: logic/alarm.php [hunk 1]
```diff
@@ -8,11 +8,16 @@
  */
 function alarm($raid, $user, $action, $info = '')
 {
+    // Get config
```
[the remark this answers is not in the document]
Comment (Chefkeks) [reply logic/alarm.php:11]: Have to disagree on this @Artanicus
Personally I prefer even such small comments, as it makes (at least for me) way easier and faster to find the spot I am looking for. But I can understand you, even though I disagree 😅

File: logic/alarm.php [hunk 2]
```diff
@@ -36,6 +41,7 @@ function alarm($raid, $user, $action, $info = '')
         $msg_text .= EMOJI_HERE . SP . $gymname . SP . '(' . $raidtimes . ')' . CR;
         $msg_text .= EMOJI_SINGLE . SP . $username . SP . '+' . $color . CR;
         $msg_text .= EMOJI_CLOCK . SP . check_time($attendtime);
+        ($config->RAID_POLL_SHOW_TRAINERCODE && $attendtime == 0 && !is_null($trainercode)) ? $msg_text .= CR . EMOJI_FRIEND . SP . $trainercode : '';
```
[the remark this answers is not in the document]
Comment (Chefkeks) [reply logic/alarm.php:44]: +1 for an own little function

File: sql/pokemon-raid-bot.sql [hunk 3]
```diff
@@ -124,6 +124,9 @@ CREATE TABLE `users` (
   `nick` varchar(100) CHARACTER SET utf8mb4 DEFAULT NULL,
   `name` varchar(200) CHARACTER SET utf8mb4 DEFAULT NULL,
   `trainername` varchar(200) CHARACTER SET utf8mb4 DEFAULT NULL,
+  `setname_time` DATETIME DEFAULT NULL,
```
Comment (Chefkeks) [inline sql/pokemon-raid-bot.sql:127]: trainername_time 😉

File: sql/pokemon-raid-bot.sql [hunk 4]
```diff
@@ -124,6 +124,9 @@ CREATE TABLE `users` (
...
+  `trainercode` varchar(12) CHARACTER SET utf8mb4 DEFAULT NULL,
+  `setcode_time` DATETIME DEFAULT NULL,
```
Comment (Chefkeks) [inline sql/pokemon-raid-bot.sql:129]: trainercode_time

**Review by Chefkeks — CHANGES_REQUESTED** [review]

**Comment by Chefkeks** [comment]
@Artanicus and @studplus
If I looked into things correctly, all changes from the trainername PR are in this one too - or not? 🤔😅

So why not closing the other PR, put all the stuff combined into this trainercode PR and then just merge this one here?

PS: Great work so far @studplus 🥳 snd don’t get all the comments wrong and change requests wrong. We just want to take the right way with the raidbot especially for new stuff as we have already a lot historically grown things 😂🙈

**Pull request closed without merging**